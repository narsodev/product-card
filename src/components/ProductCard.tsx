import React from 'react'

import { createContext, CSSProperties } from 'react'
import useProduct from '../hooks/useProduct'
import { InitialValues, onChangeArgs, Product, ProductCardHandlers } from '../interfaces/interfaces'

import { ProductContextProps } from '../interfaces/interfaces'

import styles from '../styles/styles.module.css'

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export interface Props {
  // children?: ReactElement | ReactElement[],
  className?: string,
  initialValues?: InitialValues,
  product: Product,
  style?: CSSProperties,
  value?: number,
  children: (args: ProductCardHandlers) => JSX.Element,
  onChange?: (args: onChangeArgs) => void
}


const ProductCard = ({ product, children, className = '', style, onChange, value, initialValues }: Props) => {
  const { counter, isMaxCountReached, maxCount, increaseBy, reset } = useProduct({ product, onChange, value, initialValues })

  return (
    <Provider value={{
      counter,
      isMaxCountReached,
      product,
      increaseBy
    }}>
      <div
        className={`${styles.productCard} ${className}`}
        style={style}
      >
        {children({
          count: counter,
          isMaxCountReached,
          maxCount,
          product,
          increaseBy,
          reset
        })}
      </div>
    </Provider>
  )
}

export default ProductCard
