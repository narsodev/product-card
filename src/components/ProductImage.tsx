import React from 'react'

import { CSSProperties, useContext } from 'react'
import { ProductContext } from './ProductCard'

import noImage from '../assets/no-image.jpg'

import styles from '../styles/styles.module.css'

export interface Props {
  className?: string,
  img?: string,
  style?: CSSProperties
}

const ProductImage = ({ img, className = '', style }: Props) => {
  const { product } = useContext(ProductContext)
  const imgToShow = img ?? product.img ?? noImage

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToShow}
      alt="Product"
      style={style}
    />
  )
}

export default ProductImage
