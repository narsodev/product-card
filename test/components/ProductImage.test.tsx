import React from 'react'
import renderer from 'react-test-renderer'
import ProductImage from '../../src/components/ProductImage'
import ProductCard from '../../src/components/ProductCard'
import { product2 } from '../data/products'

describe('Product Image', () => {
  it('renders with img', () => {
    const wrapper = renderer.create(
      <ProductImage img='./cofee.jpg' />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('renders with product title', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {() => <ProductImage />}
      </ProductCard>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
