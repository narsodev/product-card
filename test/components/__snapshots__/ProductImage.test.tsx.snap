// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Image renders with img 1`] = `
<img
  alt="Product"
  className="productImg "
  src="./cofee.jpg"
/>
`;

exports[`Product Image renders with product title 1`] = `
<div
  className="productCard "
>
  <img
    alt="Product"
    className="productImg "
    src="./coffee-mug.png"
  />
</div>
`;
